### PR TITLE
[generator.py] Use argparse and option for paths

### PR DESCRIPTION
**Pull request by sophie-h**

### CODE REVIEW

**Comment by bilelmoussaoui** [comment]
Overall the changes looks very good to me, mind updating the one in gtk-rs/gtk-rs as well?

**Comment by sophie-h** [comment]
Good point @bilelmoussaoui, I included the last change from gtk-rs. I will create a PR after this is approved.

File: generator.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
-#!/usr/bin/env python3
```
Comment (GuillaumeGomez) [inline generator.py:1]: Please revert this change. The path isn't always `/usr/bin` for python3 (in my case it's in `sbin` for example :p)

File: generator.py [hunk 2]
```diff
@@ -78,102 +83,72 @@ def regen_crates(path, conf, level=0):
         if isdir(entry_file):
             if level < 2 and not regen_crates(entry_file, conf, level + 1):
                 return False
-        elif entry == 'Gir.toml':
+        elif fnmatch.fnmatchcase(entry, "Gir*.toml"):
```
Comment (GuillaumeGomez) [inline generator.py:86]: Seems a bit overkill compared to just `elif entry.startwith("Gir") and entry.endswith(".toml"):`...

File: generator.py [hunk 3]
```diff
@@ -10,6 +12,9 @@
 NEED_UPDATE = 1
 FAILURE = 2
 
+DEFAULT_GIR_FILES = 'gir-files'
```
Comment (GuillaumeGomez) [inline generator.py:15]: What do you think about `DEFAULT_GIR_FILES_FOLDER` instead?

File: generator.py [hunk 4]
```diff
@@ -78,102 +83,72 @@ def regen_crates(path, conf, level=0):
...
-    if conf["gir_files"] is None:
+    if conf.gir_files == DEFAULT_GIR_FILES:
         if def_check_submodule("gir-files", conf) == FAILURE:
             return 1
-        conf["gir_files"] = "gir-files"
-    if conf["gir_path"] is None:
+    else:
+       if not isdir(conf.gir_files):
+            print("`{}` dir doesn't exist. Aborting...".format(path))
+            return 1
```
Comment (GuillaumeGomez) [inline generator.py:136]:
```suggestion
    elif not isdir(conf.gir_files):
        print("`{}` dir doesn't exist. Aborting...".format(path))
        return 1
```

File: generator.py [hunk 5]
```diff
@@ -78,102 +83,72 @@ def regen_crates(path, conf, level=0):
...
+
+    if conf.gir_path == DEFAULT_GIR_PATH:
         if not build_gir_if_needed(def_check_submodule("gir", conf)):
             return 1
-        conf["gir_path"] = "./gir/target/release/gir"
+    else:
+       if not isfile(conf.gir_path):
+            print("`{}` file doesn't exist. Aborting...".format(path))
+            return 1
```
Comment (GuillaumeGomez) [inline generator.py:144]:
```suggestion
   elif not isfile(conf.gir_path):
        print("`{}` file doesn't exist. Aborting...".format(path))
        return 1
```

File: generator.py [hunk 6]
```diff
@@ -78,102 +83,72 @@ def regen_crates(path, conf, level=0):
...
 
     print('=> Regenerating crates...')
-    if not regen_crates(".", conf):
-        return 1
-    if conf["run_fmt"] is True and not run_command(['cargo', 'fmt']):
+    for path in conf.path:
```
Comment (GuillaumeGomez) [inline generator.py:147]: Nice add!

**Review by GuillaumeGomez — CHANGES_REQUESTED** [review]
Lot of small things but overall it's pretty nice! I especially like the new `path` option.

File: generator.py (hunk 6 above)
Comment (sophie-h) [reply generator.py:147]: That's the only thing I wanted to change and I thought it would be simpler using argparse xD

File: generator.py (hunk 3 above)
Comment (sophie-h) [reply generator.py:15]: This is named after "--gir-files". Would only rename it if we rename the option.

Comment (sophie-h) [reply generator.py:15]: And `--gir-files-dir` would probably be clearer

File: generator.py (hunk 2 above)
Comment (sophie-h) [reply generator.py:86]: I would call it python style, but I don't care.

File: generator.py (hunk 4 above)
Comment (sophie-h) [reply generator.py]: Ups

File: generator.py (hunk 5 above)
Comment (sophie-h) [reply generator.py]: Fixed

File: generator.py (hunk 3 above)
Comment (GuillaumeGomez) [reply generator.py:15]: Then please do the change here and open an issue (or even a PR) on `gir`. :)

Comment (sophie-h) [reply generator.py:15]: Wait, I'm confused. What does gir have to do with this?

Comment (sophie-h) [reply generator.py:15]: If we do it consistent with `gir`, it should be `--gir-directory`

Comment (GuillaumeGomez) [reply generator.py:15]: Arf sorry, don't know why I thought you were talking about `gir`... So yes, please feel free to update the option name here.

Comment (GuillaumeGomez) [reply generator.py:15]: Fine by me!

**Comment by sophie-h** [comment]
Done from my side

File: generator.py [hunk 7]
```diff
@@ -78,102 +82,70 @@ def regen_crates(path, conf, level=0):
...
+    return parser.parse_args()
 
 
 def main():
     gir_path = None
 
     conf = parse_args(sys.argv[1:])
-    if conf is None:
-        return 1
+    print(conf)
```
Comment (GuillaumeGomez) [inline generator.py:127]: Maybe remove this line? ;)

Comment (sophie-h) [reply generator.py]: How pedantic of you :angry: :laughing:

**Review by GuillaumeGomez — APPROVED** [review]

File: generator.py (hunk 7 above)
Comment (GuillaumeGomez) [reply generator.py]: I like `print`, but too much `print` kills my brain. :p

**Comment by GuillaumeGomez** [comment]
Thanks!
